### DISCUSSION TITLE
Implementing New Platforms

### DISCUSSION THREAD

**Discussion by d-lask** [jrouwe/JoltPhysics#711] [General]
Hello,

I just started using Jolt recently and have really been enjoying it! As I start getting further in implementing it in my own project/game engine, I was curious if there was any insight on what it would take to implement or build Jolt for a new platform. I know there's support for _Platform Blue_, but I'm curious if one were to try and add other similar platforms where they might start, or rather if anyone had insight on what would be necessary/the types of pitfalls they may encounter.

I saw in the Core.h mention of needing to wrap platform Tick functions, but then I also saw a specific Mutex implementation for Platform Blue in the headers as well.

If this is something that needs investigation I'd be eager to try and give it a go or contribute if possible. Also not trying to break any NDA's, so I'm looking more for high level suggestions rather than any API specifics, etc.

Thank you for your time!

**Comment by jrouwe** [comment]
I think adding a new platform is extremely simple and shouldn't cost you more than a couple of hours. 

The mutex is overridden for Platform Blue only because ```std::mutex``` and ```std::shared_mutex``` turned out to not be very efficient under high concurrency. On all other platforms it just uses the ```std``` versions, so I wouldn't expect you to need to do anything there.

The only thing you really need to do is to is implement:
* JPH_BREAKPOINT which you can trivially copy from the x86 or ARM implementation in most cases.
* GetProcessorTickCount which can return 0 if you don't care about the profiling reports.
* You may hit one of the 9 ```#errror Undefined``` lines depending on how exotic your architecture or compiler is (e.g. you possibly have to implement functions like CountTrailingZeros, CountLeadingZeros or CountBits).

If your platform is x86 or ARM based then it will automatically use SSE/Neon intrinsics, if not it will fall back to plain C++ implementations.

B.t.w. if you need to support Platform Blue then I can provide you with PlatformBlue.h through the Platform Blue Partners website (which would prove that you're a registered developer and have signed their NDA).

**Reply by d-lask** [reply]
This is exactly what I was looking for. Thanks a bunch!

**Comment by nirvin** [comment]
@jrouwe Could I get PlatformBlue.h through the Platform Blue Partners website? Was just looking to integrate Jolt and that would be a time saver.

**Reply by jrouwe** [reply]
Hello, I just posted the file in the 'Free Talk' forum, thread 'Instructions for compiling Jolt Physics on PS4/PS5 ' with ID 324558.

(I'm assuming this is for 5 and not 4)

**Reply by nirvin** [reply]
Awesome. Thank you!

**Reply by jrouwe** [reply]
Note that I'm not in the office for the next couple of weeks, so if you post a reply there, ping me here and I'll answer your question there.
